homeassistant/components/androidtv_remote/media_player.py

```diff
@@ -0,0 +1,163 @@
+    _attr_supported_features = (
+        MediaPlayerEntityFeature.PAUSE
+        | MediaPlayerEntityFeature.VOLUME_STEP
+        | MediaPlayerEntityFeature.VOLUME_MUTE
+        | MediaPlayerEntityFeature.VOLUME_SET
+        | MediaPlayerEntityFeature.PREVIOUS_TRACK
+        | MediaPlayerEntityFeature.NEXT_TRACK
+        | MediaPlayerEntityFeature.TURN_ON
+        | MediaPlayerEntityFeature.TURN_OFF
+        | MediaPlayerEntityFeature.PLAY
+        | MediaPlayerEntityFeature.STOP
+        | MediaPlayerEntityFeature.PLAY_MEDIA
+    )
+
+    def __init__(self, api: AndroidTVRemote, config_entry: ConfigEntry) -> None:
+        """Initialize the entity."""
+        super().__init__(api, config_entry)
+
+        self._update_current_app(api.current_app)
+        self._update_volume_info(api.volume_info)
+
+        @callback
+        def current_app_updated(current_app: str) -> None:
+            self._update_current_app(current_app)
+            self.async_write_ha_state()
+
+        @callback
+        def volume_info_updated(volume_info: dict[str, str | bool]) -> None:
+            self._update_volume_info(volume_info)
+            self.async_write_ha_state()
+
+        api.add_current_app_updated_callback(current_app_updated)
+        api.add_volume_info_updated_callback(volume_info_updated)
+
+    def _update_current_app(self, current_app: str) -> None:
+        """Update current app info."""
+        self._attr_app_id = current_app
+        self._attr_app_name = KNOWN_APPS.get(current_app, current_app)
+
+    def _update_volume_info(self, volume_info: dict[str, str | bool]) -> None:
+        """Update volume info."""
+        self._attr_volume_level = int(volume_info["level"]) / 100
+        self._attr_is_volume_muted = bool(volume_info["muted"])
+
+    @property
+    def state(self) -> MediaPlayerState:
+        """Return the state of the device."""
+        if self._attr_is_on:
+            return MediaPlayerState.ON
+        return MediaPlayerState.OFF
+
+    async def async_turn_on(self) -> None:
+        """Turn the Android TV on."""
+        if not self._attr_is_on:
+            self._send_key_command("POWER")
+
+    async def async_turn_off(self) -> None:
+        """Turn the Android TV off."""
+        if self._attr_is_on:
+            self._send_key_command("POWER")
+
+    async def async_volume_up(self) -> None:
+        """Turn volume up for media player."""
+        self._send_key_command("VOLUME_UP")
+
+    async def async_volume_down(self) -> None:
+        """Turn volume up for media player."""
+        self._send_key_command("VOLUME_DOWN")
+
+    async def async_mute_volume(self, mute: bool) -> None:
+        """Mute the volume."""
+        if mute != self.is_volume_muted:
+            self._send_key_command("VOLUME_MUTE")
+
+    async def async_set_volume_level(self, volume: float) -> None:
+        """Set volume level, range 0..1."""
+        if isinstance(self._attr_volume_level, float):
```

Comment:
Isn't this going to throw not defined if this function is called before we get the volume_info_updated callback?

Reply:
Since now we have a floating `VOLUME_SET` Home Assistant will throw an error that the service is not supported if there is no information about the volume.